main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,396 @@
+
+    case Expression.PutChannel(exp1, exp2, tpe, loc) =>
+      val (e1, o1) = visitExp(exp1)
+      val (e2, o2) = visitExp(exp2)
+      val o3 = combineAll(o1, o2)
+      (OccurrenceAst.Expression.PutChannel(e1, e2, tpe, loc), o3)
+
+    case Expression.SelectChannel(rules, default, tpe, loc) =>
+      var rs: List[OccurrenceAst.SelectChannelRule] = List.empty
+      var o1: Map[Symbol.VarSym, Occur] = Map.empty
+      for (r <- rules) {
+        val (c, o2) = visitExp(r.chan)
+        val (e, o3) = visitExp(r.exp)
+        rs = rs :+ OccurrenceAst.SelectChannelRule(r.sym, c, e)
+        o1 = combineAll(o1, combineAll(o2, o3))
+      }
+
+      val (d, o4) = resolveOption(default, (None, o1), (x:LiftedAst.Expression) => {
+        val (d, o5) = visitExp(x)
+        (Some(d), combineAll(o5, o1))
+      })
+      (OccurrenceAst.Expression.SelectChannel(rs, d, tpe, loc), o4)
+
+    case Expression.Spawn(exp, tpe, loc) =>
+      val (e, o1) = visitExp(exp)
+      (OccurrenceAst.Expression.Spawn(e, tpe, loc), o1)
+
+    case Expression.Lazy(exp, tpe, loc) =>
+      val (e, o1) = visitExp(exp)
+      (OccurrenceAst.Expression.Lazy(e, tpe, loc), o1)
+
+    case Expression.Force(exp, tpe, loc) =>
+      val (e, o1) = visitExp(exp)
+      (OccurrenceAst.Expression.Force(e, tpe, loc), o1)
+
+    case Expression.HoleError(sym, tpe, loc) =>
+      (OccurrenceAst.Expression.HoleError(sym, tpe, loc), Map.empty)
+
+    case Expression.MatchError(tpe, loc) =>
+      (OccurrenceAst.Expression.MatchError(tpe, loc), Map.empty)
+  }
+
+  /**
+   * Performs occurrence analysis on a list of expressions 'exps' and merges occurrences
+   */
+  private def visitExps(exps: List[LiftedAst.Expression]): (List[OccurrenceAst.Expression], Map[Symbol.VarSym, Occur]) = {
+    exps.foldLeft((List[OccurrenceAst.Expression](), Map[Symbol.VarSym, OccurrenceAst.Occur]()))((acc, arg) => {
+      val (e, o1) = visitExp(arg)
+      val o2 = combineAll(o1, acc._2)
+      (acc._1 :+ e, o2)
+    })
+  }
+
+  /**
+   * Combines the 2 maps `m1` and `m2` of the type (Symbol -> Occur) into a single map of same type using the function `combine`.
+   */
+  private def combineAll(m1: Map[Symbol.VarSym, Occur], m2: Map[Symbol.VarSym, Occur]): Map[Symbol.VarSym, Occur]
+  =
+  {
+    var m3 = m1.map {
+      case (sym, o1) =>
+        val o3 = resolveOption(m2.get(sym), o1, combine(o1, _))
+        (sym, o3)
+    }
+
+    m2.foreach {
+      case (sym, o) => if (!m1.contains(sym)) m3 += (sym -> o)
+    }
+    m3
+  }
+
+  /**
+   * Combines two occurrences `o1` and `o2` of type Occur into a single occurrence.
+   */
+  private def combine(o1: Occur, o2: Occur): Occur = (o1, o2) match {
+    case (Dead, _) => o2
+    case (_, Dead) => o1
+    case _ => Many
+  }
+
+  /**
+   * Matches on `opt`. If None returns `default`, else apply `f` on value in Some(B)
+   */
+  private def resolveOption[A, B](opt: Option[B], default: A, f: B => A): A = {
```

Comment:
OK, and now you can realize it is a function in the standard library. In Scala its `.map(...).getOrElse(...)`.